Aruco Markers

ryanvade opened the pull request.

jsoutherland commented on include/wallaby/aruco.hpp:
```diff
@@ -0,0 +1,69 @@
+
+namespace aruco {
+  class Aruco;
+  static Aruco* instance;
```
We usually hide this instance variable in the ``Aruco::instance()`` function like in
https://github.com/kipr/libwallaby/blob/master/src/wallaby_p.cpp#L114-L118

jsoutherland commented:
@ryanvade I tried calibrate with this code (saved as ``calib.cpp``):
````c++
#include <wallaby/wallaby.hpp>
#include <wallaby/wallaby.h>
#include <wallaby/aruco.hpp>


#include <iostream>
#include <vector>


int main(int argc, char ** argv)
{
	std::cout << "Opening camera..." << std::endl;
	Camera::Device cam;
	cam.open(0, MED_RES, BLACK_2017);

	std::cout << "Waiting 5 seconds before first image" << std::endl;
	msleep(5000);

	const int numImgs = 30;

	std::vector<cv::Mat > images;

	for (int i = 0; i < numImgs; ++i)
	{
		cam.update();
		images.push_back(cam.rawImage());
		std::cout << (i+1) << "/" << numImgs << " images gathered" << std::endl;
		msleep(500);
	}

	cam.close();

	std::cout << "Calibrating with " << images.size() << " images" << std::endl;

	aruco::Aruco * aruco = aruco::Aruco::getInstance();

	aruco->calibrate(images);

	std::cout << "Done." << std::endl;

	return 0;
}
````

compiled as 
``g++ -o calibrate --std=c++11 calib.cpp -lwallaby -lopencv_core``

and get the error:
````
OpenCV Error: Assertion failed (nimages > 0) in calibrateCamera, file /home/kipr/opencv/modules/calib3d/src/calibration.cpp, line 3357
terminate called after throwing an instance of 'cv::Exception'
  what():  /home/kipr/opencv/modules/calib3d/src/calibration.cpp:3357: error: (-215) nimages > 0 in function calibrateCamera
````

...even though there are 30 images in the vector being passed to ``Aruco::calibrate()``

ryanvade commented:
Are you checking if the frame has the chessboard? 
```
vector<Vec2f> foundPoints;
Mat frame;
int chessBoardFlags = cv::CALIB_CB_ADAPTIVE_THRESH |cv::CALIB_CB_NORMALIZE_IMAGE | cv::CALIB_CB_FAST_CHECK;
for(int i = 0; i < numImgs; ++i)
{
    cam.Update();
    frame = cam.rawImage();
    if(findChessBoardCorners(frame, Size(6, 9), foundPoints, chessBoardFlags))
    {
        images.push_back(frame);
	std::cout << (i+1) << "/" << numImgs << " images gathered" << std::endl;
	msleep(500);
    }
}
```

jsoutherland commented:
@ryanvade Whoops, no I wasn't.   I was hoping to get enough in 30 frames but I guess something went wrong.  Good catch.

ryanvade commented:
Yeah the calibration function does not like frames without the chessboard in it.

jsoutherland commented:
@ryanvade I think aruco.cpp should be using the camera device singleton.   The new calibration function is working but it always uses the ``WHITE_CAMERA`` filename, even though I had the black camera open because ``aruco::Aruco`` is creating a new ``Camera::Device``.  I think there will be other issues that show up besides that and using the device singleton should help with the multiple ``Camera::Device::open()`` attempt issue that will probably happen again.

Another thing that will be needed to bring this in is to ``#ifdef ARUCO`` the new aruco code in libwallaby and add something like this to ``CMakeLists.txt``:
````cmake

option(USE_ARUCO "Use Aruco Markers" OFF)

if(USE_ARUCO)
	add_definitions(-DARUCO)
endif()
````
We will build without aruco support by default for the wallaby controller, for now.

ryanvade commented:
@jsoutherland  I can switch it over to the singleton. Can you link to an example of using the DeviceSingleton?

jsoutherland commented:
@ryanvade Cool, thanks.   

Here is an example:
https://github.com/kipr/libwallaby/blob/master/src/camera_c.cpp#L47

It is in the ``Private`` namespace (see: https://github.com/kipr/libwallaby/blob/master/src/camera_c_p.hpp) 

so it would be something like:
````c++
bool ret = Private::DeviceSingleton::instance()->open(number, res, model);
````

jsoutherland commented:
It is safer to re-use that instance() function throughout your Aruco class as needed, rather than trying to store a pointer as a class variable.

ryanvade commented:
Do I need to get the number, resolution, and model?

jsoutherland commented:
Sorry I was just grabbing and example using the instance, but it was misleading since you use ``open()`` as well and in a different (better) way.

I think it is probably best to pass no arguments from Aruco, so that the default args in camera.cpp are used. Like:
````c++
bool ret = Private::DeviceSingleton::instance()->open();
````

ryanvade commented:
@jsoutherland I switched everything over to the devicesingleton instance.

jsoutherland commented:
I found that msleep(500) was about 10 times too long...

ryanvade commented:
How about that last commit?

As for the calibrate function in the c library, I don't remember why I added initially but since calibration should be done before running any user program including it makes little sense. I removed that stub.

ryanvade commented:
Updates @jsoutherland ?

jsoutherland commented:
@ryanvade Sorry for the delay, I will give it a try.

jsoutherland commented:
Looks good, I hit two errors:
At  https://github.com/ryanvade/libwallaby/blob/aruco/CMakeLists.txt#L132
linking opencv_aruco should be removed from that command for a following
````cmake
if(USE_ARUCO)
    target_link_libraries(wallaby opencv_aruco)
endif()
````

The second one was my bad... would you mind updating this header path at the same time?
https://github.com/ryanvade/libwallaby/blob/aruco/src/channel_p.cpp#L10
to ``#include <opencv2/core/core.hpp>``

ryanvade commented:
Changes committed.

jsoutherland commented:
@ryanvade  Thanks, this is a nice contribution!
